### DISCUSSION TITLE
How MeterProvider works between API and SDK

### DISCUSSION THREAD

**Discussion by Deniz** [open-telemetry/opentelemetry-dotnet#4082] [Q&A] [answered]
Hi All,

I have been reading the source code of runtime (tag v7.0.2) `System.Diagnostics.DiagnosticSource` and opentelemetry dotnet client (tag 1.4.0-rc.1) in order to see how they work together. In the `System.Diagnostics.DiagnosticSource` while I was looking into the `Instrument` class I saw the use of the method which simply notifies the listener.

```
protected void RecordMeasurement(T measurement, ReadOnlySpan<KeyValuePair<string, object?>> tags)
        {
            DiagNode<ListenerSubscription>? current = _subscriptions.First;
            while (current is not null)
            {
                current.Value.Listener.NotifyMeasurement(this, measurement, tags, current.Value.State);
                current = current.Next;
            }
        }
``` 

In the listener, all callbacks are no-op which is expected I think (API should be separated from SDK, so no data should be passed)?

```
        // We initialize all measurement callback with no-op operations so we'll avoid the null checks during the execution;
        private MeasurementCallback<byte>    _byteMeasurementCallback    = (instrument, measurement, tags, state) => { /* no-op */ };
        private MeasurementCallback<short>   _shortMeasurementCallback   = (instrument, measurement, tags, state) => { /* no-op */ };
        private MeasurementCallback<int>     _intMeasurementCallback     = (instrument, measurement, tags, state) => { /* no-op */ };
        private MeasurementCallback<long>    _longMeasurementCallback    = (instrument, measurement, tags, state) => { /* no-op */ };
        private MeasurementCallback<float>   _floatMeasurementCallback   = (instrument, measurement, tags, state) => { /* no-op */ };
        private MeasurementCallback<double>  _doubleMeasurementCallback  = (instrument, measurement, tags, state) => { /* no-op */ };
        private MeasurementCallback<decimal> _decimalMeasurementCallback = (instrument, measurement, tags, state) => { /* no-op */ };
```

Correct me if I am wrong. Now, the real question here is - actually two questions - 

1) There is a `AggregationManager` class which sets the callbacks in order to update the `InstrumentState` and again in that `AggregationManager` class depending on the type of the Instrument, an aggregator is being assigned in order to be called upon new measurement inside the `InstrumentState`. I wonder who is the consumer of that `AggregationManager` and how it's being used?

2) In the SDK code, also `MeterProviderSdk` uses a different aggregator (`AggregatorStore`) via `Metric` class. My question here is, how the code in `System.Diagnostics.DiagnosticSource` interacts with the `MeterProvider` in the SDK. 
When we include the SDK and configure it, do we replace the logic in the `System.Diagnostics.DiagnosticSource` with the one in the SDK by updating callbacks in `MeterProviderSdk` so `MeterListener` belongs to it will be responsible for collecting the data rather than the one in the API?

In summary, how the data pass between API and SDK parts, which gives us the Metric to be exported?

Thanks,

**Comment by cijothomas** [comment, marked as answer]
1. `AggregationManager` is not at all connected/related to OpenTelemetry. Its used by .NET runtime itself for tools like dotnet-counters.

2. OpenTelemetry uses MeterListener to subscribe to Metric updates from the API (DiagnosticSource): https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry/Metrics/MeterProviderSdk.cs#L152
There is no logic to be replaced - DS gives call backs. OTel uses it to get raw measurements, and aggregates them following OTel spec.

**Reply by Deniz** [reply]
Thanks @cijothomas, so all aggregation happens in the SDK part? If `AggregationManager` is not connected to OTel, collecting data from `Instrument` defined in DiagnosticSource & aggregation happens in the dotnet-client (SDK) by using `MeterListener` and using `AggregatorStore`? You mean `InstrumentState` and other things I mentioned from `DiagnosticSource` is irrelevant to OTel (SDK)? 

`AggregationManager` confused me, sorry if question is not clear.

**Reply by cijothomas** [reply]
All aggregation is done inside OTel SDK.
